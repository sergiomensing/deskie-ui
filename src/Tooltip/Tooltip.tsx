import { Children, PropsWithChildren, useState, HTMLAttributes, cloneElement } from 'react';
import { usePopper } from 'react-popper';
import styled from 'styled-components';

type Placement = 'top' | 'bottom' | 'right' | 'left';

const StyledTooltip = styled.div<{ isOpen: boolean; placement: Placement }>`
  .tooltip {
    display: flex;
    flex-direction: column;
    background-color: #11142d;
    color: ${(props) => props.theme.colors.labelInverted};
    padding: 0.5rem;
    border-radius: 0.5rem;
    box-shadow: 0 0.3px 0.3px -4px rgba(0, 0, 0, 0.081), 0 1.1px 2px -4px rgba(0, 0, 0, 0.119),
      0 5px 10px -4px rgba(0, 0, 0, 0.2);

    transform: scale(0.8);
    transform-origin: center;
    opacity: 0;
    transition: 0.2s cubic-bezier(0.45, 0, 0.21, 1);

    ${(props) =>
      props.isOpen &&
      `
    transform: scale(1);
    opacity: 1;
  `}
  }
`;

type TooltipProps = HTMLAttributes<unknown> & {
  content: React.ReactNode;
  placement?: Placement;
};

function Tooltip({
  children,
  content,
  placement = 'bottom',
  className = '',
  ...rest
}: PropsWithChildren<TooltipProps>) {
  const [isOpen, setIsOpen] = useState<boolean>(false);

  const [referenceElement, setReferenceElement] = useState<HTMLDivElement | null>(null);
  const [popperElement, setPopperElement] = useState<HTMLDivElement | null>(null);

  const { styles, attributes } = usePopper(referenceElement, popperElement, {
    placement: placement,
    modifiers: [
      {
        name: 'offset',
        options: {
          offset: [0, 4],
        },
      },
    ],
  });

  const renderTarget = () => {
    const target = Children.only(children);

    // @ts-ignore
    return cloneElement(target, {
      ref: setReferenceElement,
      onMouseEnter: () => window.innerWidth >= 1024 && setIsOpen(true),
    });
  };

  return (
    <StyledTooltip
      placement={placement}
      isOpen={isOpen}
      className={`tooltip-container ${className}`}
      {...rest}
      onMouseLeave={() => setIsOpen(false)}
    >
      {renderTarget()}
      <div
        ref={setPopperElement}
        style={{ ...styles.popper, zIndex: 99, pointerEvents: 'none' }}
        {...attributes.popper}
      >
        <div className="tooltip">{content}</div>
      </div>
    </StyledTooltip>
  );
}

export default Tooltip;
