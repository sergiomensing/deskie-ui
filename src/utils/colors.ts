export const colorShade = (col: string, amt: number) => {
  col = col.replace(/^#/, '');
  if (col.length === 3) col = col[0] + col[0] + col[1] + col[1] + col[2] + col[2];

  const [sr, sg, sb] = col.match(/.{2}/g) || ['ff', 'ff', 'ff'];
  const [ar, ag, ab] = [parseInt(sr, 16) + amt, parseInt(sg, 16) + amt, parseInt(sb, 16) + amt];

  const r = Math.max(Math.min(255, ar), 0).toString(16);
  const g = Math.max(Math.min(255, ag), 0).toString(16);
  const b = Math.max(Math.min(255, ab), 0).toString(16);

  const rr = (r.length < 2 ? '0' : '') + r;
  const gg = (g.length < 2 ? '0' : '') + g;
  const bb = (b.length < 2 ? '0' : '') + b;

  return `#${rr}${gg}${bb}`;
};

const defaultColors = [
  '#E64980',
  '#4C6EF5',
  '#228BE6',
  '#15AABF',
  '#12B886',
  '#40C057',
  '#82C91E',
  '#FAB005',
  '#FD7E14',
  '#FA5252',
];

export function getColorFromString(value: string, colors: undefined | string[] = defaultColors) {
  let hash = 0;
  if (value.length === 0) return colors[hash];
  for (let i = 0; i < value.length; i++) {
    hash = value.charCodeAt(i) + ((hash << 5) - hash);
    hash = hash & hash;
  }
  hash = ((hash % colors.length) + colors.length) % colors.length;
  return colors[hash];
}
