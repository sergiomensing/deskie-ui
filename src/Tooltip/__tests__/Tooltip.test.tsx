import React from 'react';
import Tooltip from '../Tooltip';
import { fireEvent, render, act, waitFor } from '../../../tests/utils';

const renderTooltip = async () => {
  let result;
  await act(async () => {
    result = await render(
      <Tooltip content="test-value">
        <span>test-value-target</span>
      </Tooltip>,
    );
  });
  return result;
};

describe('Tooltip', () => {
  it('should render', async () => {
    const { getByText, container } = await renderTooltip();

    await waitFor(() => {
      expect(getByText('test-value')).toBeInTheDocument();
      expect(getByText('test-value-target')).toBeInTheDocument();

      expect(container).toMatchSnapshot();
    });
  });

  it('should appear on hover target element', async () => {
    const { getByText, container } = await renderTooltip();

    await waitFor(() => {
      expect(window.getComputedStyle(container.querySelector('.tooltip')).opacity).toBe('0');
    });

    act(() => {
      fireEvent.mouseEnter(getByText('test-value-target'));
    });

    await waitFor(() => {
      expect(window.getComputedStyle(container.querySelector('.tooltip')).opacity).toBe('1');
    });
  });
});
