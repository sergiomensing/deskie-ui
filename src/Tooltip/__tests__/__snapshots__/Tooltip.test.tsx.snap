// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should render 1`] = `
.c0 .tooltip {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background-color: #11142d;
  color: #FFFFFF;
  padding: 0.5rem;
  border-radius: 0.5rem;
  box-shadow: 0 0.3px 0.3px -4px rgba(0,0,0,0.081),0 1.1px 2px -4px rgba(0,0,0,0.119),0 5px 10px -4px rgba(0,0,0,0.2);
  -webkit-transform: scale(0.8);
  -ms-transform: scale(0.8);
  transform: scale(0.8);
  -webkit-transform-origin: center;
  -ms-transform-origin: center;
  transform-origin: center;
  opacity: 0;
  -webkit-transition: 0.2s cubic-bezier(0.45,0,0.21,1);
  transition: 0.2s cubic-bezier(0.45,0,0.21,1);
}

<div>
  <div
    class="c0 tooltip-container "
  >
    <span>
      test-value-target
    </span>
    <div
      data-popper-escaped="true"
      data-popper-placement="bottom"
      data-popper-reference-hidden="true"
      style="position: absolute; left: 0px; top: 0px; z-index: 99; pointer-events: none; transform: translate(0px, 4px);"
    >
      <div
        class="tooltip"
      >
        test-value
      </div>
    </div>
  </div>
</div>
`;
